File: backend/app/dcs/pip/provider.py
from sqlalchemy.orm import Session
from fastapi import Request
from app.core.config import settings
from app.core.security.auth import Principal
from app.dcs.pep.mode import dcs_enabled
from app.dcs.pip.types import Subject, Context, Resource, PolicyInput
from app.dcs.pip.classification import get_classification_map
from app.observability.perf import perf_span

def build_policy_input(
    *,
    db: Session,
    request: Request,
    principal: Principal,
    action: str,
    resource_type: str,
    resource_id: str,
    owner_id: str | None,
    labels: list[str] | None = None,
    crypto_meta: dict[str, dict] | None = None,
) -> PolicyInput:
    with perf_span("pip_ms"):
        labels = labels or []
        crypto_meta = crypto_meta or {}

        subject = Subject(
            user_id=principal.user_id,
            tenant_id=principal.tenant_id,
            role=principal.role,
            username=principal.username,
        )

        ctx = Context(
            env=settings.ENV,
            channel="web",
            purpose="cinema_ops",
            client_ip=request.headers.get("x-real-ip") or (request.client.host if request.client else None),
            device_trust=0.8,  # PoC constant
            request_id=getattr(request.state, "request_id", "no-request-id"),
        )

        if not dcs_enabled():
            res = Resource(
                type=resource_type,
                id=resource_id,
                owner_id=owner_id,
                tenant_id=principal.tenant_id,
                labels=labels,
                fields={},
            )
            return PolicyInput(subject=subject, action=action, resource=res, context=ctx)

        cls_map = get_classification_map(db, resource_type)

        fields: dict[str, dict] = {}
        for field_name, classification in cls_map.items():
            fields[field_name] = {"classification": classification}
            if field_name in crypto_meta:
                fields[field_name]["crypto"] = crypto_meta[field_name]

        res = Resource(
            type=resource_type,
            id=resource_id,
            owner_id=owner_id,
            tenant_id=principal.tenant_id,
            labels=labels,
            fields=fields,
        )

        return PolicyInput(subject=subject, action=action, resource=res, context=ctx)

File: backend-go/internal/dcs/pip/provider.go
package pip

import (
	"context"

	"github.com/neoweyss/poc-dcs/backend-go/internal/dcs/cache"
	"github.com/neoweyss/poc-dcs/backend-go/internal/dcs/runtime"
	"github.com/neoweyss/poc-dcs/backend-go/internal/dcs/types"
	"github.com/neoweyss/poc-dcs/backend-go/internal/observability/perf"
)

type ClassificationStore interface {
	GetByResourceType(ctx context.Context, resourceType string) (map[string]types.Classification, error)
}

type Config struct {
	Env            string
	Channel        string
	Purpose        string
	DeviceTrust    float64
	ClientIPHeader string
}

type Input struct {
	Principal    types.Principal
	Action       string
	ResourceType string
	ResourceID   string
	OwnerID      string
	Labels       []string
	Request      types.RequestContext
	CryptoMeta   map[string]map[string]string
}

type Provider struct {
	runtime *runtime.Settings
	cache   *cache.Manager
	store   ClassificationStore
	cfg     Config
}

func NewProvider(rt *runtime.Settings, cm *cache.Manager, store ClassificationStore, cfg Config) *Provider {
	return &Provider{
		runtime: rt,
		cache:   cm,
		store:   store,
		cfg:     cfg,
	}
}

func (p *Provider) Build(ctx context.Context, in Input) (types.PolicyInput, error) {
	defer perf.Span(ctx, "pip_ms")()

	fields := map[string]types.FieldMeta{}
	if p.runtime.DcsEnabled() {
		clsMap, err := p.getClassificationMap(ctx, in.ResourceType)
		if err != nil {
			return types.PolicyInput{}, err
		}
		for fieldName, cls := range clsMap {
			meta := types.FieldMeta{Classification: cls}
			if crypto, ok := in.CryptoMeta[fieldName]; ok {
				meta.Crypto = crypto
			}
			fields[fieldName] = meta
		}
	}

	reqCtx := in.Request
	if reqCtx.Channel == "" {
		reqCtx.Channel = p.cfg.Channel
	}
	if reqCtx.Purpose == "" {
		reqCtx.Purpose = p.cfg.Purpose
	}
	if reqCtx.Env == "" {
		reqCtx.Env = p.cfg.Env
	}
	if reqCtx.DeviceTrust == 0 {
		reqCtx.DeviceTrust = p.cfg.DeviceTrust
	}

	return types.PolicyInput{
		Principal: in.Principal,
		Action:    in.Action,
		Resource: types.Resource{
			Type:     in.ResourceType,
			ID:       in.ResourceID,
			OwnerID:  in.OwnerID,
			TenantID: in.Principal.TenantID,
			Labels:   in.Labels,
			Fields:   fields,
		},
		Context: reqCtx,
	}, nil
}

func (p *Provider) getClassificationMap(ctx context.Context, resourceType string) (map[string]types.Classification, error) {
	if p.cache.LevelEnabled(1) {
		if cached, ok := p.cache.Classification.Get(resourceType); ok {
			return cached, nil
		}
	}

	rows, err := p.store.GetByResourceType(ctx, resourceType)
	if err != nil {
		return nil, err
	}
	if p.cache.LevelEnabled(1) {
		p.cache.Classification.Set(resourceType, rows, 0)
	}
	return rows, nil
}
